lib/ssz_ex.ex

```diff
@@ -8,22 +8,222 @@ defmodule LambdaEthereumConsensus.SszEx do
+    with {:ok, {encoded_list, byte_size_list, total_byte_size}} <-
+           Enum.reduce_while(list, {:ok, {[], [], 0}}, fn value,
+                                                          {:ok, {res_encoded, res_size, acc}} ->
+             case encode(value, basic_type) do
+               {:ok, encoded} ->
+                 size = byte_size(encoded)
+                 {:cont, {:ok, {[encoded | res_encoded], [size | res_size], size + acc}}}
+
+               error ->
+                 {:halt, {:error, error}}
+             end
+           end) do
+      {:ok, {Enum.reverse(encoded_list), Enum.reverse(byte_size_list), total_byte_size}}
+    end
+  end
+
+  defp decode_list(binary, basic_type, size) do
+    fixed_size = get_fixed_size(basic_type)
+
+    with {:ok, decoded_list} = result <-
+           binary
+           |> decode_chunk(fixed_size, basic_type)
+           |> flatten_results() do
+      if length(decoded_list) > size do
+        {:error, "invalid max_size of list"}
+      else
+        result
+      end
+    end
+  end
+
+  defp decode_variable_list(binary, _, _) when byte_size(binary) == 0 do
+    {:error, "Error trying to collect empty list"}
+  end
+
+  defp decode_variable_list(binary, basic_type, size) do
+    <<first_offset::integer-32-little, rest_bytes::bitstring>> = binary
+    num_items = div(first_offset, @bytes_per_length_offset)
+
+    if Integer.mod(first_offset, @bytes_per_length_offset) != 0 ||
+         first_offset < @bytes_per_length_offset do
+      {:error, "InvalidListFixedBytesLen"}
+    else
+      with {:ok, first_offset} <-
+             sanitize_offset(first_offset, nil, byte_size(binary), first_offset) do
+        {:ok, {decoded_list, _, _}} =
+          decode_variable_list(first_offset, rest_bytes, num_items, binary, basic_type)
+
+        if length(decoded_list) > size do
+          {:error, "invalid length list"}
+        else
+          {:ok, decoded_list |> Enum.reverse()}
+        end
+      end
+    end
+  end
+
+  @spec decode_variable_list(integer(), binary, integer(), binary, any) ::
+          {:ok, {list, integer(), binary}} | {:error, String.t()}
+  defp decode_variable_list(first_offset, rest_bytes, num_items, binary, basic_type) do
+    1..num_items
+    |> Enum.reduce_while({:ok, {[], first_offset, rest_bytes}}, fn i,
+                                                                   {:ok,
+                                                                    {acc_decoded, offset,
+                                                                     acc_rest_bytes}} ->
+      if i == num_items do
+        part = :binary.part(binary, offset, byte_size(binary) - offset)
+
+        with {:ok, decoded} <- decode(part, basic_type) do
+          {:cont, {:ok, {[decoded | acc_decoded], offset, rest_bytes}}}
+        end
+      else
+        get_next_offset(acc_decoded, acc_rest_bytes, basic_type, offset, binary, first_offset)
+      end
+    end)
```

Comment:
I find this part difficult to understand. Also, the inner `with` has the same problem from `get_next_offset`: it can return an `{:error, error}` tuple that needs handling. I think we should take an approach similar to `decode_chunk` on this one.